modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/mapping/MappingServiceImpl.java

```diff
@@ -258,4 +271,37 @@ private static List<Fragment> replace(
 
         return newFragments;
     }
+
+    /**
+     * Holder for topology snapshots that guarantees monotonically increasing versions.
+     */
+    class LogicalTopologyHolder {
+        private volatile List<String> nodes = List.of();
```

Comment:
why volatile used ?

Reply:
> why volatile used ?

Topology changes leads write to the field inside a synchronized block,
however, mapper service reads the field without obtaining a lock.

> i.e. can we obtain situation when onNodeLeft and map are called in parallel ?

yes